File: Src2/Icm.Core/Repository/MemoryRepositoryOfTEntityTKey.vb
Imports Icm.Reflection
Imports System.Linq.Expressions

Namespace Icm.Data
    Public Class MemoryRepository(Of TEntity, TKey)
        Implements IRepository(Of TEntity, TKey)

        Protected store As ICollection(Of TEntity)
        Private ReadOnly _idFunction As Expression(Of Func(Of TEntity, TKey))

        ''' <summary>
        ''' Initializes a new instance of the MemoryRepository class.
        ''' </summary>
        ''' <param name="idFunction"></param>
        Public Sub New(ByVal idFunction As Expression(Of Func(Of TEntity, TKey)))
            store = New HashSet(Of TEntity)
            _idFunction = idFunction
        End Sub

        ''' <summary>
        ''' Initializes a new instance of the MemoryRepository class.
        ''' </summary>
        ''' <param name="store"></param>
        ''' <param name="idFunction"></param>
        Public Sub New(ByVal store As ICollection(Of TEntity), ByVal idFunction As Expression(Of Func(Of TEntity, TKey)))
            Me.store = store
            _idFunction = idFunction
        End Sub

        Public Sub Load(entities As IEnumerable(Of TEntity))
            store = New HashSet(Of TEntity)(entities)
        End Sub

        Public Sub Add(entity As TEntity) Implements IRepository(Of TEntity, TKey).Add
            store.Add(entity)
        End Sub

        Public Sub Delete(entity As TEntity) Implements IRepository(Of TEntity, TKey).Delete
            store.Remove(entity)
        End Sub

        Private Function IdEqualsExpression(key As TKey) As Expression(Of Func(Of TEntity, Boolean))
            ' Going from _idFunction, that has the form "Function(x) x.Id",
            ' to "Function(x) x.Id = VALUE" (where VALUE is the supplied key)

            ' So first we build the equality expression "x.Id = VALUE":
            Dim equalExpr = Expression.Equal(_idFunction.Body, Expression.Constant(key))

            ' Then we create a new lambda with the equality expression as body and the same parameter of _idFunction.
            Return Expression.Lambda(Of Func(Of TEntity, Boolean))(equalExpr, _idFunction.Parameters.First)
        End Function

        Public Overridable Function GetById(key As TKey) As TEntity Implements IRepository(Of TEntity, TKey).GetById
            Return store.AsQueryable().SingleOrDefault(IdEqualsExpression(key))
        End Function

        Public Function GetEnumerator() As IEnumerator(Of TEntity) Implements IEnumerable(Of TEntity).GetEnumerator
            Return store.GetEnumerator
        End Function

        Public Function GetEnumerator1() As IEnumerator Implements IEnumerable.GetEnumerator
            Return store.GetEnumerator
        End Function

        Public Function Contains(entity As TEntity) As Boolean Implements IRepository(Of TEntity, TKey).Contains
            Return store.Contains(entity)
        End Function

        Public Function ContainsId(key As TKey) As Boolean Implements IRepository(Of TEntity, TKey).ContainsId
            Return store.AsQueryable().Any(IdEqualsExpression(key))
        End Function
    End Class

End Namespace

File: Src/Icm.Core/Repository/MemoryRepositoryOfTEntityTKey.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Icm.Reflection;
using System.Linq.Expressions;

namespace Icm.Data
{
	public class MemoryRepository<TEntity, TKey> : IRepository<TEntity, TKey>
	{

		protected ICollection<TEntity> store;

		private readonly Expression<Func<TEntity, TKey>> _idFunction;
		/// <summary>
		/// Initializes a new instance of the MemoryRepository class.
		/// </summary>
		/// <param name="idFunction"></param>
		public MemoryRepository(Expression<Func<TEntity, TKey>> idFunction)
		{
			store = new HashSet<TEntity>();
			_idFunction = idFunction;
		}

		/// <summary>
		/// Initializes a new instance of the MemoryRepository class.
		/// </summary>
		/// <param name="store"></param>
		/// <param name="idFunction"></param>
		public MemoryRepository(ICollection<TEntity> store, Expression<Func<TEntity, TKey>> idFunction)
		{
			this.store = store;
			_idFunction = idFunction;
		}

		public void Load(IEnumerable<TEntity> entities)
		{
			store = new HashSet<TEntity>(entities);
		}

		public void Add(TEntity entity)
		{
			store.Add(entity);
		}

		public void Delete(TEntity entity)
		{
			store.Remove(entity);
		}

		private Expression<Func<TEntity, bool>> IdEqualsExpression(TKey key)
		{
			// Going from _idFunction, that has the form "Function(x) x.Id",
			// to "Function(x) x.Id = VALUE" (where VALUE is the supplied key)

			// So first we build the equality expression "x.Id = VALUE":
			dynamic equalExpr = Expression.Equal(_idFunction.Body, Expression.Constant(key));

			// Then we create a new lambda with the equality expression as body and the same parameter of _idFunction.
			return Expression.Lambda<Func<TEntity, bool>>(equalExpr, _idFunction.Parameters.First);
		}

		public virtual TEntity GetById(TKey key)
		{
			return store.AsQueryable().SingleOrDefault(IdEqualsExpression(key));
		}

		public IEnumerator<TEntity> GetEnumerator()
		{
			return store.GetEnumerator;
		}

		public IEnumerator GetEnumerator1()
		{
			return store.GetEnumerator;
		}
		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator1();
		}

		public bool Contains(TEntity entity)
		{
			return store.Contains(entity);
		}

		public bool ContainsId(TKey key)
		{
			return store.AsQueryable().Any(IdEqualsExpression(key));
		}
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
